### PR TITLE
NPC weapon switching improvement + UPS bug fix

### PR DESCRIPTION
**Pull request by Coolthulhu**
Changes:
- NPCs consider UPS charges when selecting ranged weapons with UPS cost
- NPCs will now take ranged weapons if they have enough ammo
- Fixed debugmsg spam if a NPC wields an UPS ranged weapon and attempts to use it
- Allow NPCs to reload gunmods
- Rewritten `gun_value`

NPCs will not wield a gun that can only be used because of the gunmod, though.

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -2001,3 +2001,36 @@ bool Character::pour_into( vehicle &veh, item &liquid )
     }
     return true;
 }
+
+long Character::ammo_count_for( const item &gun )
+{
+    // INT_MAX instead of LONG_MAX because some code may not expect > INT_MAX
```
Comment (mugling) [inline src/character.cpp]: That's probably true; do you know which code specifically?

File: src/character.cpp [hunk 2]
```diff
@@ -2001,3 +2001,36 @@ bool Character::pour_into( vehicle &veh, item &liquid )
...
+    long ret = INT_MAX;
+    if( !gun.is_gun() ) {
+        return ret;
+    }
+
+    long required = gun.ammo_required();
+
+    if( required > 0 ) {
+        long remaining = 0;
+        if( gun.magazine_integral() ) {
```
Comment (mugling) [inline src/character.cpp]: Why the check for integral magazines?

File: src/character.cpp (hunk 1 above)
Comment (Coolthulhu) [inline src/character.cpp]: I don't know which one specifically, but I recall there being a bunch of `ints` around.
Java/C# got it right with numeric values requiring explicit casting.

Comment (codemime) [inline src/character.cpp]: I believe it has to be a [named constant](https://github.com/CleverRaven/Cataclysm-DDA/pull/16901/files#diff-d13ed72411aea98b2ed21e2cb17387ebR572) everywhere.

File: src/character.cpp (hunk 2 above)
Comment (Coolthulhu) [inline src/character.cpp]: Kinda forgot what I was doing here, I guess.

I want this section to:
* Always add `ammo_remaining`
* Only find nearby ammo if at least one usable magazine is found
* Get the total number of shots that can be fired with ammo in gun + usable ammo nearby (if any)

**Comment by Coolthulhu** [comment]
Ran into a small problem: rating weapons by ammo count for them makes the NPC prefer swapping guns to full ones (if they're good enough) instead of emptying one and then swapping.
Not a big deal since it doesn't happen all that often, but it makes them act less human-like.

File: src/ranged.cpp [hunk 3]
```diff
@@ -1667,3 +1669,117 @@ void drop_or_embed_projectile( const dealt_projectile_attack &attack )
     }
 }
 
+double player::gun_value( const item &weap, long ammo ) const
+{
+    // TODO: Mods
+    // TODO: Allow using a specified type of ammo rather than default
+    if( weap.type->gun.get() == nullptr ) {
```
Comment (BevapDin) [inline src/ranged.cpp]: Can't you just check `item::is_gun`?

File: src/ranged.cpp [hunk 4]
```diff
@@ -1667,3 +1669,117 @@ void drop_or_embed_projectile( const dealt_projectile_attack &attack )
...
+        return 0.0;
+    }
+
+    if( ammo <= 0 ) {
+        return 0.0;
+    }
+
+    const islot_gun& gun = *weap.type->gun.get();
```
Comment (BevapDin) [inline src/ranged.cpp]: Calling `get()` isn't needed, `unique_ptr` has an overload `operator*`, so `*weap.type->gun` is enough.

**Pull request merged**